Add documentation of how to write an actor to upgrade process. (#500)
oamg/leapp: docs/source/actor-rhel7-o-rhel8.md

# How to create a Leapp actor for RHEL 7 to 8 upgrade

## Introduction

This document is intended for all people who want to contribute to RHEL upgrade process using Leapp tool. After reading through this document, you will be able to transform your expertise in certain parts of RHEL into improvements of the RHEL 7 to 8 upgrade tooling. You are expected to be familiar with the concept of in-place upgrades and ideally to have at least high-level knowledge about the Leapp tool.

## Setting up the development environment

> Leapp actors are written in Python, so your usual Python development setup can be used during the process of creating a new actor.

Reviewer: This may raise a question which exactly python. Should we specify that code has to be compatible with both py2 and py3?

## Tools

The main tools you will use for the actor development are listed below.

### leapp

The leapp framework uses provides the libraries required to be imported by any actor and also a binary tool used to control the execution of actors within a workflow.  

> ### snactor

Reviewer: As a newcomer I might like to see the source code for the tool, can you add a link?

    },
    "type": "Report"
  }
]
```

To flush all saved messages from the repository database, run `snactor messages clear`.

## Writing tests for an actor

[Read the tutorial](https://leapp.readthedocs.io/en/latest/unit-testing.html) for writing an running unit and component tests

## Best practices

Read the best practices [document](https://github.com/oamg/leapp-repository/blob/master/docs/best-practices.md)

## Contributing actors to the Leapp project

Currently all Leapp elements (i.e. actors, models, tags) are stored under a public [GitHub repository](https://github.com/oamg/leapp-repository)

All new content that needs to be part of Leapp release distributed to all users should be proposed as a Pull Request in this repository.

**Before submitting your work for review, make sure you have read and followed the contribution guidelines:**

[Contribution guidelines for writing actors](https://github.com/oamg/leapp-repository/blob/master/CONTRIBUTING.md)

> ## FAQ

Reviewer: Maybe it's not a bad idea to give an example of "what we expect" or an ideal PR. Luckily we have such an example https://github.com/oamg/leapp-repository/pull/186/

### What Python version my actor/tests code should be compatible with?

Python 2.7+/3.6+.

### How to add tests to my new actor?

Under “tests” folder, an actor can have Python files containing tests that will be executed using PyTest. Leapp provide tests utilities to simulate actor consuming and checking actor production. Please, refer to detailed Leapp documentation about how to write tests.

### How to use libraries or data files in my new actor?

An actor can have data files under the “files” folder and Python libraries under the “libraries” folder. Leapp does provide utilities to help an actor to access the files and libraries in these folders. Please, refer to detailed Leapp documentation about this.

### Where can I report an issue or RFE related to the framework or other actors?

You can open a GitHub issues:

- [Leapp framework](https://github.com/oamg/leapp/issues/new)
- [Leapp actors](https://github.com/oamg/leapp-repository/issues/new)

When filing an issue, include:

- How to reproduce it
- All files in `/var/log/leapp`.
- The `/var/lib/leapp/leapp.db` file.

## Where can I seek help?

> We’ll gladly answer your questions and lead you to through any troubles with the actor development. You can reach us, the OS and Application Modernization Group, by these mailing list: leapp-devel@redhat.com or at freenode IRC server in channel __#leapp__.

Reviewer 2: ```suggestion
We’ll gladly answer your questions and lead you to through any troubles with the actor development. You can reach us, the OS and Application Modernization Group, at freenode IRC server in channel __#leapp__.
```

# How to create a Leapp actor for RHEL 7 to 8 upgrade

## Introduction

This document is intended for all people who want to contribute to RHEL upgrade process using Leapp tool. After reading through this document, you will be able to transform your expertise in certain parts of RHEL into improvements of the RHEL 7 to 8 upgrade tooling. You are expected to be familiar with the concept of in-place upgrades and ideally to have at least high-level knowledge about the Leapp tool.

## Setting up the development environment

> Leapp actors are written in Python 2.7+/3.6+, so your usual Python development setup can be used during the process of creating a new actor.

Reviewer 2: I'm not sure if it's just me but this implies that it could be either 2.7 or 3.6 rather than it should be compatible with both

Reviewer: Same here, maybe adding second sentence with precise requirements will help? Something like
"But keep in mind that the resulting code has to be both py2 and py3 compatible" ?

[
  {
    "stamp": "2019-04-30T13:12:05.706317Z",
    "hostname": "leapp-20190429150826",
    "actor": "check_os_release",
    "topic": "report_topic",
    "context": "0ac49430-1b29-4940-92bb-3e81da85f8af",
    "phase": "NON-WORKFLOW-EXECUTION",
    "message": {
      "hash": "686355c648832f5399cffc5bd0e7e8312d2b77577cdeec94f674a2565090cf7f",
      "data": "{\"audience\": [\"sysadmin\"], \"detail\": \"{\\\"summary\\\": \\\"Minimal supported OS version for upgrade process: 7.6\\\"}\", \"flags\": [\"inhibitor\"], \"renderers\": {\"html\": \"<h2 class=\\\"report-title\\\">{{ title }}</h2><p class=\\\"report-summary\\\">{{ summary }}</p>\", \"plaintext\": \"{{ title }}\\n{{ summary }}\"}, \"severity\": \"medium\", \"title\": \"Unsupported OS version\"}"
    },
    "type": "Report"
  }
]
```

To flush all saved messages from the repository database, run `snactor messages clear`.

## Writing tests for an actor

[Read the tutorial](https://leapp.readthedocs.io/en/latest/unit-testing.html) for writing an running unit and component tests

## Best practices

> Read the best practices [document](https://github.com/oamg/leapp-repository/blob/master/docs/best-practices.md)

Reviewer: Please add a link to the general 'python guidelines' document https://github.com/oamg/leapp-guidelines/blob/master/python-coding-guidelines.md
We should merge the best practices and guidelines one day as well.

## Contributing actors to the Leapp project

Currently all Leapp elements (i.e. actors, models, tags) are stored under a public [GitHub repository](https://github.com/oamg/leapp-repository)

All new content that needs to be part of Leapp release distributed to all users should be proposed as a Pull Request in this repository.

**Before submitting your work for review, make sure you have read and followed the contribution guidelines:**

[Contribution guidelines for writing actors](https://github.com/oamg/leapp-repository/blob/master/CONTRIBUTING.md)

> You can see this [pull request](https://github.com/oamg/leapp-repository/pull/186) as example.

Reviewer: I would emphasize why this PR is perfect. 
"This [PR](link) gives a good example of both guidelines-driven actor implementation and thorough test coverage"

# How to create a Leapp actor for RHEL 7 to 8 upgrade

## Introduction

> This document is intended for all people who want to contribute to RHEL upgrade process using Leapp tool. After reading through this document, you will be able to transform your expertise in certain parts of RHEL into improvements of the RHEL 7 to 8 upgrade tooling. You are expected to be familiar with the concept of in-place upgrades and ideally to have at least high-level knowledge about the Leapp tool.

Reviewer 3: ```suggestion
This document is intended for all people who want to contribute to the process of upgrading Red Hat Enterprise Linux (RHEL) 7 to RHEL 8 using Leapp tool. The upgrade is performed in place meaning that the RHEL 7 installation is replaced by RHEL 8 on the same storage.
After reading through this document, you will be able to transform your expertise in certain parts of RHEL into improvements of the RHEL 7 to 8 upgrade tooling.
```

# cd /usr/share/leapp-repository/repositories/system_upgrade/el7toel8/actors/mynewactor/
# tree
.
├── actor.py
└── tests
```

The main file of the actor is the actor.py in which you’ll write code for logic of the actor.

For further information about how create an actor read this [document](https://leapp.readthedocs.io/en/latest/first-actor.html)

## Including an actor in the RHEL 7 to 8 upgrade process

Until now, you have created boilerplate of a new actor and made it visible to Leapp. But, Leapp needs some more information about what to do with the actor. Specifically, in which **“workflow”** and in which **“phase”** the actor should be executed. A workflow is a sequence of phases. The only workflow available now is the one solving the upgrade of RHEL 7 to RHEL 8. Each phase is a set of actors that will be executed one after another before the next phase starts. To find out in which workflow and phase should the actor be executed, Leapp looks for **“tags”**. To be part of RHEL 7 to RHEL 8 upgrade workflow, an actor needs to be tagged with **IPUWorkflowTag**.

> The phases of the IPUWorkflow (in order) are: **Facts Collection, Checks, Report, Download, Upgrade RamDisk Preparation, Upgrade RamDisk Start, Late Tests, Preparation, RPM Upgrade, Application Upgrade, Third Party Applications, Finalization, Reboot** and **First Boot**. Each phase has a specific tag that marks an actor as being part of that phase. You can find descriptions of all the phases and their tags [here](https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/workflows/inplace_upgrade.py) and [here](https://leapp-to.github.io/upgrades).

Reviewer 2: ```suggestion
The phases of the IPUWorkflow (in order) are: **Facts Collection, Checks, Report, Download, Upgrade RamDisk Preparation, Upgrade RamDisk Start, Late Tests, Preparation, RPM Upgrade, Application Upgrade, Third Party Applications, Finalization, Reboot** and **First Boot**. Each phase has a specific tag that marks an actor as being part of that phase. You can find descriptions of all the phases and their tags [here](https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/workflows/inplace_upgrade.py).
```

## Contributing actors to the Leapp project

Currently all Leapp elements (i.e. actors, models, tags) are stored under a public [GitHub repository](https://github.com/oamg/leapp-repository)

All new content that needs to be part of Leapp release distributed to all users should be proposed as a Pull Request in this repository.

**Before submitting your work for review, make sure you have read and followed the contribution guidelines:**

[Contribution guidelines for writing actors](https://github.com/oamg/leapp-repository/blob/master/CONTRIBUTING.md)

This [pull request](https://github.com/oamg/leapp-repository/pull/186) gives a good example of both guidelines-driven actor implementation and thorough test coverage

## Pull requests review processes

Every pull request need to be reviewed by core-team members, for optimize this process please ensure your code is following some criteria:

- Remove [wip][poc][tbd][donotmerge] comments and labels. WIP pull requests won't be reviewed unless specifically asked by the PR author.

> - Make sure you have covered a new actor with unit tests and these are passing if they don't and you have no idea why please seek help/guidance.

Reviewer: ```suggestion
- Make sure you have covered a new actor with unit tests and these are passing, if they don't and you have no idea why please seek help/guidance.
```

> - Test your implemented functionality doing the upgrade process before open a PR.

Reviewer: ```suggestion
- Test your implemented functionality by doing the upgrade process before opening a PR.
```

> - We usually take at least 3 work days to review if you need some priority in this process please let us know.

Reviewer: ```suggestion
- We usually take at least 3 working days to review a new PR, if you need some priority in this process please let us know.
```

> - All criteria from the [best practices](https://github.com/oamg/leapp-repository/blob/master/docs/best-practices.md) and [Python guidelines](https://github.com/oamg/leapp-guidelines/blob/master/python-coding-guidelines.md) are expected.

Reviewer: ```suggestion
- We expect the code to meet [leapp best practices](https://github.com/oamg/leapp-repository/blob/master/docs/best-practices.md) and [leapp python guidelines](https://github.com/oamg/leapp-guidelines/blob/master/python-coding-guidelines.md).
```

## FAQ

### In which existing workflow phase should I place my new actor?

> You can decide that based on the description of the [phases](https://leapp-to.github.io/upgrades). Please note that if your actor depends on some message generated by another actor, it cannot be executed in a phase before the phase of such actor. In a similar way, if your actor produces data, it needs to be executed before the actor consuming the data.

Reviewer 4: Obsoleted link to leapp-to IO pages

Reviewer: Unless we have a more up-to-date information about phases I'd rather keep a link to the old one than no link at all. It gives at least some idea what to grep for in the codebase.

Reviewer 2: Phases are listed higher (L64) with link to code containing description.
Phases in the github.io page do not match what we have in repository and will only confuse people, so I would get rid of it.
